consumer/consumer.go

```diff
@@ -0,0 +1,192 @@
+package consumer
+
+import (
+	"fmt"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/Shopify/sarama"
+	cluster "github.com/bsm/sarama-cluster"
+	"github.com/heetch/felice/consumer/handler"
+	"github.com/heetch/felice/message"
+	"github.com/pkg/errors"
+)
+
+// Consumer is a Kafka consumer.
+type Consumer struct {
+	consumer      *cluster.Consumer
+	config        *cluster.Config
+	handlers      *handler.Collection
+	wg            sync.WaitGroup
+	quit          chan struct{}
+	RetryInterval time.Duration
+	metrics       MetricsHook
+}
+
+// Handle registers the handler for the given topic.
+// Handle must be called before Serve.
+func (c *Consumer) Handle(topic string, h handler.Handler) {
+	c.setup()
+
+	c.handlers.Set(topic, h)
+}
+
+func (c *Consumer) setup() {
+	if c.handlers == nil {
+		c.handlers = &handler.Collection{}
+	}
+
+	if c.quit == nil {
+		c.quit = make(chan struct{})
+	}
+
+	if c.RetryInterval == 0 {
+		c.RetryInterval = time.Second
+	}
+}
+
+// SetMetricsHook registers a MetricsHook on the Consumer that will be invoked with datas about the handled messages.
+func (c *Consumer) SetMetricsHook(mh MetricsHook) {
```

Comment:
What about exporting the `metrics` field (l24) as we did for the `RetryInterval` field? The `setup` function will then make sure these fields have default values when running the producer. cf https://middlemost.com/object-lifecycle/#please-just-use-fields

Reply:
And what will be the default value for the `Metrics` field?

Reply:
If the zero value is fine, this is your default value, you don't need to manually set it